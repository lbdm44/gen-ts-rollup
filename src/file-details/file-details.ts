export class FileDetails {
  defaultExport: string | undefined;

  namedExports: string[] = [];

  constructor(public filePath: string) {}
}
