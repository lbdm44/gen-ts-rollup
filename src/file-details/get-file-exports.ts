import { parse } from '@babel/parser';
import traverse from '@babel/traverse';
import * as fs from 'fs-extra';

import type { FileDetails } from './file-details';
import { debug } from '../utils/debug';

const getFileExportsDebug = debug.extend('get-file-exports');

export interface FileExports {
  defaultExport?: string;

  namedExports: string[];
}

export function getFileExports(fileDetails: FileDetails): FileExports {
  const { filePath } = fileDetails;
  const fileContents = fs.readFileSync(filePath, 'utf8');

  const fileExports: FileExports = {
    namedExports: [],
  };

  const ast = parse(fileContents, {
    sourceType: 'module',
    plugins: [
      'decorators-legacy',
      'classProperties',
      'classPrivateProperties',
      'typescript',
    ],
  });

  traverse(ast, {
    ExportDefaultDeclaration: (declPath) => {
      const {
        node: { declaration },
      } = declPath;

      // if (declaration.type === 'CallExpression') {
      //   // TODO: This pattern is rather brittle, essentially `export default foo()` probsbly should do `const f = foo(); export default f;`
      //   fileExports.defaultExport = declaration.callee?.name;
      // }

      if (declaration.type === 'Identifier') {
        fileExports.defaultExport = declaration.name;
      } else if (
        declaration.type === 'ClassDeclaration' ||
        declaration.type === 'FunctionDeclaration'
      ) {
        // Only supported named classes and functions at this time.
        fileExports.defaultExport = declaration.id?.name;
      } else {
        getFileExportsDebug(
          `Default Export Node Type Not Handled: ${declPath.node?.declaration?.type} in ${filePath}`
        );
      }
    },
    ExportNamedDeclaration: (declPath) => {
      const {
        node: { declaration, specifiers },
      } = declPath;

      /**
       * TODO: handle this scenario
       *
       * `export { default } from 'foo/bar';`
       *
       * Ends up being `import { default } from '../../foo/bar'`
       */

      if (specifiers.length) {
        specifiers.forEach((specifier) => {
          if (
            specifier.type === 'ExportDefaultSpecifier' ||
            specifier.type === 'ExportNamespaceSpecifier'
          ) {
            fileExports.namedExports.push(specifier.exported.name);
          } else if (specifier.exported.type === 'Identifier') {
            fileExports.namedExports.push(specifier.exported.name);
          } else {
            fileExports.namedExports.push(specifier.exported.value);
          }
        });
      } else if (
        declaration?.type === 'ClassDeclaration' ||
        declaration?.type === 'TSEnumDeclaration' ||
        declaration?.type === 'TSInterfaceDeclaration' ||
        declaration?.type === 'TSTypeAliasDeclaration'
      ) {
        fileExports.namedExports.push(declaration.id.name);
      } else if (
        declaration?.type === 'FunctionDeclaration' &&
        declaration.id
      ) {
        // Does not currently support anonymous functions
        fileExports.namedExports.push(declaration.id.name);
      } else if (declaration?.type === 'VariableDeclaration') {
        declaration.declarations.forEach((varDecl) => {
          if (varDecl.id.type === 'Identifier') {
            fileExports.namedExports.push(varDecl.id.name);
          }
        });
      } else {
        getFileExportsDebug(
          `Named export: ${declaration?.type} not handled in ${filePath}`
        );
      }
    },
  });

  return fileExports;
}
