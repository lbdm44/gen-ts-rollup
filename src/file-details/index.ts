export { FileDetails } from './file-details';

export { getFileExports } from './get-file-exports';
