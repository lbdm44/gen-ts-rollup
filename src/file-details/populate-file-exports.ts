import { FileDetails } from './file-details';
import { getFileExports } from './get-file-exports';

export function populateFileExports(fileDetails: FileDetails): FileDetails {
  const fileExports = getFileExports(fileDetails);

  fileDetails.defaultExport = fileExports.defaultExport;
  fileDetails.namedExports = [...fileExports.namedExports];

  return fileDetails;
}
