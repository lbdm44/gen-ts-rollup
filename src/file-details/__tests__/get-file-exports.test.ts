import * as path from 'path';
import { describe, expect, test } from 'vitest';
import { FileDetails } from '../file-details';
import { getFileExports } from '../get-file-exports';

describe('get-file-exports', () => {
  const FIXTURES_PATH = path.resolve(__dirname, './__fixtures__');

  describe('js files', () => {
    test('it should get exports for a simple JS file', () => {
      const jsFilePath = path.resolve(FIXTURES_PATH, 'js-file.js');
      const fileDetails = new FileDetails(jsFilePath);
      const fileExports = getFileExports(fileDetails);

      expect(fileExports.defaultExport).toBeUndefined();
      expect(fileExports.namedExports).toEqual(['add']);
    });

    test('it should handle default class exports', () => {
      const jsFilePath = path.resolve(
        FIXTURES_PATH,
        'js-files/default-class-export.js'
      );
      const fileDetails = new FileDetails(jsFilePath);
      const fileExports = getFileExports(fileDetails);

      expect(fileExports.defaultExport).toBe('Foo');
      expect(fileExports.namedExports).toHaveLength(0);
    });

    test('it should handle default function exports', () => {
      const jsFilePath = path.resolve(
        FIXTURES_PATH,
        'js-files/default-fn-export.js'
      );
      const fileDetails = new FileDetails(jsFilePath);
      const fileExports = getFileExports(fileDetails);

      expect(fileExports.defaultExport).toBe('foo');
      expect(fileExports.namedExports).toHaveLength(0);
    });

    test('it should handle named class exports', () => {
      const jsFilePath = path.resolve(
        FIXTURES_PATH,
        'js-files/class-export.js'
      );
      const fileDetails = new FileDetails(jsFilePath);
      const fileExports = getFileExports(fileDetails);

      expect(fileExports.defaultExport).toBeUndefined();
      expect(fileExports.namedExports).toMatchSnapshot();
    });

    test('it should handle named function exports', () => {
      const jsFilePath = path.resolve(FIXTURES_PATH, 'js-files/fn-export.js');
      const fileDetails = new FileDetails(jsFilePath);
      const fileExports = getFileExports(fileDetails);

      expect(fileExports.defaultExport).toBeUndefined();
      expect(fileExports.namedExports).toMatchSnapshot();
    });

    test('it should handle complex exports', () => {
      const jsFilePath = path.resolve(
        FIXTURES_PATH,
        'js-files/complex-exports.js'
      );
      const fileDetails = new FileDetails(jsFilePath);
      const fileExports = getFileExports(fileDetails);

      expect(fileExports.defaultExport).toBe('foo');
      expect(fileExports.namedExports).toMatchSnapshot();
    });
  });

  describe('ts files', () => {
    const TS_FILES_PATH = path.resolve(FIXTURES_PATH, 'ts-files');

    test('it should get exports for a simple TS file', () => {
      const tsFile = path.resolve(FIXTURES_PATH, 'ts-file.ts');
      const fileDetails = new FileDetails(tsFile);

      const fileExports = getFileExports(fileDetails);

      expect(fileExports.defaultExport).toBeUndefined();
      expect(fileExports.namedExports).toEqual(['add']);
    });

    test('it should handle default class exports', () => {
      const tsFilePath = path.resolve(TS_FILES_PATH, 'default-class-export.ts');
      const fileDetails = new FileDetails(tsFilePath);
      const fileExports = getFileExports(fileDetails);

      expect(fileExports.defaultExport).toBe('Foo');
      expect(fileExports.namedExports).toHaveLength(0);
    });

    test('it should handle default function exports', () => {
      const tsFilePath = path.resolve(TS_FILES_PATH, 'default-fn-export.ts');
      const fileDetails = new FileDetails(tsFilePath);
      const fileExports = getFileExports(fileDetails);

      expect(fileExports.defaultExport).toBe('foo');
      expect(fileExports.namedExports).toHaveLength(0);
    });

    test('it should handle complex exports', () => {
      const tsFilePath = path.resolve(
        FIXTURES_PATH,
        'ts-files/complex-exports.ts'
      );
      const fileDetails = new FileDetails(tsFilePath);
      const fileExports = getFileExports(fileDetails);

      expect(fileExports.defaultExport).toBe('foo');
      expect(fileExports.namedExports).toMatchSnapshot();
    });
  });
});
