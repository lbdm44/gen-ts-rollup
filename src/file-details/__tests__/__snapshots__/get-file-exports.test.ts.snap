// Vitest Snapshot v1

exports[`get-file-exports > js files > it should handle complex exports 1`] = `
[
  "Foo",
  "foo",
  "BAR",
  "baz",
]
`;

exports[`get-file-exports > js files > it should handle named class exports 1`] = `
[
  "Foo",
  "Bar",
]
`;

exports[`get-file-exports > js files > it should handle named function exports 1`] = `
[
  "foo",
  "bar",
]
`;

exports[`get-file-exports > ts files > it should handle complex exports 1`] = `
[
  "Foo",
  "foo",
  "BAR",
  "baz",
]
`;
